text/0738-ember-data-deprecate-model-reopen.md

```diff
@@ -0,0 +1,57 @@
+---
+Stage: Accepted
+Start Date: 2021-04-23
+Release Date: Unreleased
+Release Versions:
+  ember-source: vX.Y.Z
+  ember-data: vX.Y.Z
+Relevant Team(s): ember-data
+RFC PR: https://github.com/emberjs/rfcs/pull/738
+---
+
+# EmberData | Deprecate Model Reopen
+
+## Summary
+
+Deprecates using `reopen` to alter clases extending `@ember-data/model`.
+
+## Motivation
+
+`reopen` restricts the ability of EmberData to design better primitives that maintain
+compatibility with `@ember-data/model`, and is a footgun that leads to confusing incorrect
+states for users that utilize it.
+
+## Detailed design
+
+The static `reopen` method on `Model` will be overwritten to provide a deprecation which
+once resolved or after the deprecation lifecycle completes will result in `reopen` throwing
+an error when used in dev and silently no-oping in production. This deprecation will target
+`5.0` and not be `enabled` sooner than `4.1` though it may come available before that.
+
+## How we teach this
+
+This is best taught through a deprecation guide. Users using `reopen` to test multiple
+configurations in their test suite should instead extend and register a new model each time.
+
+Users using `reopen` to modify a class immediately after creating it should also refactor
+to `extend` instead.
+
+Users using `reopen` to modify a class dynamically at runtime should refactor to either register
+new model types or (better) utilize a megamorphic solution such as ember-m3 to achieve their needs.
+
+In all cases, using `reopen` after a class instance for a record has already been created has *always*
+resulted in at least minor and potentially major errors in application state.
```

Comment:
Basically there are three issues though I don't think they need to be laid out here in the rfc text itself:

1) is that we cache the metas for relationships and attributes, so if a reopen were to redefine a relationship (or add a potential resolvable inverse to a relationship where none was before) then those definitions would likely not get applied to new instances as expected (we'd use old metas anywhere we pulled from cache which is most places).

2) is that existing instantiated relationships don't get reconfigured which means that if their definition has been changed they are now potentially a significantly different shape.

3) is a subset of (1). Serialization and normalization typically breaks because eachAttribute/eachRelationship will often be missing necessary definitions.